File: data/cpp/tests/test_Whoosh_terms.cpp
#include <gtest/gtest.h>

#include "../src/function_Whoosh_terms.cpp"

class TestWhooshListTerms : public ::testing::Test {
protected:
    std::string index_dir = "test_list_terms_index";
    WhooshTerms func;

    void SetUp() override {
        cleanup();
        func.manage_index_and_list_terms<std::string>(index_dir);
    }

    void TearDown() override {
        cleanup();
    }

    void cleanup() {
        if (std::filesystem::exists(index_dir)) {
            std::filesystem::remove_all(index_dir);
        }
    }
};

TEST_F(TestWhooshListTerms, test_list_terms) {
    std::string terms = func.manage_index_and_list_terms<std::string>(index_dir);
    EXPECT_TRUE(terms.find("hello") != std::string::npos);
    EXPECT_TRUE(terms.find("test") != std::string::npos);
    EXPECT_GT(terms.size(), 0);
}

TEST_F(TestWhooshListTerms, test_invalid_index_directory) {
    std::string invalid_index_dir = "invalid_index";
    if (std::filesystem::exists(invalid_index_dir)) {
        std::filesystem::remove_all(invalid_index_dir);
    }
    EXPECT_THROW(func.manage_index_and_list_terms<std::string>(invalid_index_dir, "content", false), std::runtime_error);
}

TEST_F(TestWhooshListTerms, test_case_insensitivity) {
    std::string terms = func.manage_index_and_list_terms<std::string>(index_dir);
    EXPECT_TRUE(terms.find("whoosh") != std::string::npos);
    EXPECT_TRUE(terms.find("Whoosh") == std::string::npos);
}

TEST_F(TestWhooshListTerms, test_word_frequency) {
    std::string terms = func.manage_index_and_list_terms<std::string>(index_dir);
    EXPECT_GT(terms.size(), 0);
}

TEST_F(TestWhooshListTerms, test_term_existence) {
    std::string terms = func.manage_index_and_list_terms<std::string>(index_dir);
    EXPECT_TRUE(terms.find("whoosh") != std::string::npos);
    EXPECT_TRUE(terms.find("nonexistentword") == std::string::npos);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: data/python/tests/test_Whoosh_terms.py
import unittest
import shutil
import os
from function_Whoosh_terms import WhooshTerms

class TestWhooshListTerms(unittest.TestCase):

    def setUp(self):
        self.index_dir = "test_list_terms_index"
        self.func = WhooshTerms()
        self.func.manage_index_and_list_terms(self.index_dir)

    def tearDown(self):
        if os.path.exists(self.index_dir):
            shutil.rmtree(self.index_dir, ignore_errors=True)

    def test_list_terms(self):
        terms = self.func.manage_index_and_list_terms(self.index_dir)
        self.assertIn("hello", terms)
        self.assertIn("test", terms)
        self.assertTrue(len(terms) > 0)

    def test_invalid_index_directory(self):
        invalid_index_dir = "invalid_index"
        if os.path.exists(invalid_index_dir):
            shutil.rmtree(invalid_index_dir)
        with self.assertRaises(ValueError):
            self.func.manage_index_and_list_terms(invalid_index_dir, add_documents=False)
            
    def test_case_insensitivity(self):
        terms = self.func.manage_index_and_list_terms(self.index_dir)
        self.assertIn("whoosh", terms)
        self.assertNotIn("Whoosh", terms)

    def test_word_frequency(self):
        terms = self.func.manage_index_and_list_terms(self.index_dir)
        self.assertGreater(len(terms), 0)

    def test_term_existence(self):
        terms = self.func.manage_index_and_list_terms(self.index_dir)
        self.assertIn("whoosh", terms)
        self.assertNotIn("nonexistentword", terms)

if __name__ == '__main__':
    unittest.main()
